### PR TITLE
Extract intensities

### LINKED ISSUE #37: Add replace argument to extract_bin_files
*This is for internal use only; if you'd like to open an issue or request a new feature, please open a bug or enhancement issue*

**Section 1: Design details**

**Relevant background**

Addresses #29.

Currently, when specifying channels to generate intensity images for, the remaining channels have blank intensity tifs saved. 

**Design overview**

We would like to add a `replace=True` default argument to `extract_bin_files()` to combine the separate pulse and intensity dimensions into a single column to write images for. In addition, the area (intensity x width) image data will be removed from both `extract_bin_files()` and `_write_out()`.

**Cases:**

1. intensities are extracted and replace the pulse data for specific channels 
2. intensities are extracted and have their own dimension 
3. intensities are not extracted

For cases 1 and 3, the img_data will only have one dimension for the type. For case 2, the implementation does not change (and so the blank tif issue still needs to be addressed). 

**Design list/flowchart**

```
def extract_bin_files():
    for i, (fov, bf) in enumerate(bin_files):
        img_data = _extract_bin.c_extract_bin(
            bytes(bf, 'utf-8'), fov['lower_tof_range'],
            fov['upper_tof_range'], np.array(fov['calc_intensity'], dtype=np.uint8)
        )

### new implementation

        if type_utils.any_true(intensities) and replace:
            if type(intensities) is not list:
                intensities = fov['targets']
            for j, target in enumerate(intensities):
                img_data[0, :, :, j] = img_data[1, :, :, j]
            img_data = img_data[0, :, :, :]
        elif not intensities:
            img_data = img_data[0, :, :, :]

###

        if out_dir is not None:
            _write_out(img_data, out_dir, fov['bin'][:-4], fov['targets'], intensities)
        else:

### adjustment for return data
            if replace or not intensities: 
                type_list = ['pulse']
            else: 
                type_list = ['pulse', 'intensities']

            image_data.append(
                xr.DataArray(
                    data=img_data[np.newaxis, :],
                    coords=[
                        [fov['bin'].split('.')[0]],
                        type_list,
                        np.arange(img_data.shape[1]),
                        np.arange(img_data.shape[2]),
                        list(fov['targets']),
                    ],
                    dims=['fov', 'type', 'x', 'y', 'channel'],
                    )   
                 )
```

The `_write_out()` function will need to be adjusted as well.  We'll pass the intensities argument as either a bool or list as defined in `extract_bin()` (rather than passing `any_true(intensities)`) which will specify which intensity tifs to save if needed.

```
def _write_out():
    out_dirs = [
        os.path.join(out_dir, fov_name),
        os.path.join(out_dir, fov_name, 'intensities'),
    ]
    suffixes = [
        '',
        '_intensity',
    ]
    save_dtypes = [
        np.uint32,
        np.uint32,
    ]

    for i, (out_dir_i, suffix, save_dtype) in enumerate(zip(out_dirs, suffixes, save_dtypes)):

### rather than relying on `any_true(intensities)` bool a before, will depend on shape of (collapsed or non-collapsed) img_data to know whether to continue loop
        if i > img_data.shape[0]:
            break
        if not os.path.exists(out_dir_i):
            os.makedirs(out_dir_i)        
        for j, target in enumerate(targets):

### should solve blank intensity tif issue by only saving images for channels specified
            if i == 0 or target in intensities:
                io.imsave(
                    os.path.join(out_dir_i, f'{target}{suffix}.tiff'),
                    img_data[i, :, :, j].astype(save_dtype),
                    plugin='tifffile',
                    check_contrast=False
                )           
```

There are multiple subdirectories created to separately hold the pulse, intensity, and area images, but with replace=True we will simply keep all extracted images in a single directory. So `_write_out()` will only create an intensities subdirectory if replace is false.

**Section 2: Implementation details**

Once you have completed section 1, please tag the relevant parties and iterate on the initial design details until everyone is satisfied. Then, 
proceed to section 2

**Control flow**

Provide additional, more granular details (if necessary) about how the proposed coding logic will be put together

**Milestones and timeline**

List each of the major components of the project, and provide an estimated completion date for each one.

### PR DESCRIPTION
**Pull request by camisowers**
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

Closes #37 

**Remaining issues**

If there are specific areas of your implementation that you feel need specific feedback, or if there are things you aren't sure about, please highlight them here

### CODE REVIEW

File: mibi_bin_tools/bin_files.py [hunk 1]
```diff
@@ -79,30 +79,32 @@ def _write_out(img_data: np.ndarray, out_dir: str, fov_name: str, targets: List[
...
         np.uint32,
     ]
+
+    if not type_utils.any_true(intensities):
+        intensities = [intensities]
+
     for i, (out_dir_i, suffix, save_dtype) in enumerate(zip(out_dirs, suffixes, save_dtypes)):
-        if i > 0 and not intensities:
-            continue
+        if i+1 > img_data.shape[0]:
```
Comment (ngreenwald) [inline mibi_bin_tools/bin_files.py:96]: Here and below as well

File: mibi_bin_tools/bin_files.py [hunk 2]
```diff
@@ -323,21 +329,40 @@ def extract_bin_files(data_dir: str, out_dir: Union[str, None],
             bytes(bf, 'utf-8'), fov['lower_tof_range'],
             fov['upper_tof_range'], np.array(fov['calc_intensity'], dtype=np.uint8)
         )
+
+        if type_utils.any_true(intensities):
+            if type(intensities) is not list:
+                intensities = fov['targets']
```
Comment (ngreenwald) [inline mibi_bin_tools/bin_files.py:335]: Here as well

File: tests/bin_files_test.py [hunk 3]
```diff
@@ -80,49 +80,73 @@ def case_specified_intensities(self):
...
+    img_data_ext = np.zeros((2, 10, 10, 5), dtype=np.uint32)
     fov_name = 'fov1'
     targets = [chr(ord('a') + i) for i in range(5)]
+    intensities = [chr(ord('a') + i) for i in range(3)]
 
     with tempfile.TemporaryDirectory() as tmpdir:
-        # correctness
-        bin_files._write_out(img_data, tmpdir, fov_name, targets)
-        filepath_checks(tmpdir, fov_name, targets, True)
+        # correct write out without intensities
```
Comment (ngreenwald) [inline tests/bin_files_test.py:137]: Now that we're substituting one file type (intensity image) for another (pulse count), it's not enough to just check that the file is created in the right place. We also need to make sure that the right file is actually being saved. I would recommend mocking `_extract_bin.c_extract_bin` to return an array with known values at each index. Then, you can check to make sure the right data is being saved to the right spot

Comment (ngreenwald) [reply tests/bin_files_test.py:137]: Now that I've written all of that out, it probably makes sense to factor our the dimensionality merging code into a helper function. That way the correctness checks can happen just on that function

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good. However, now that the logic is more complex, the testing is going to be more complex as well.

File: tests/bin_files_test.py (hunk 3 above)
Comment (camisowers) [reply tests/bin_files_test.py:137]: sounds good

**Review by ngreenwald — APPROVED** [review]
Looks good! We'll need a new tag/release so that we can bump the import in toffy

**Pull request merged**